File: backend/hackathon-backend/routers/audit.py
"""
Router for audit log-related endpoints
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from database import get_db
from schemas import AuditLog, AuditLogCreate
from crud import (
    create_audit_log, get_audit_logs, get_audit_logs_by_entity
)

router = APIRouter()

@router.get("/", response_model=List[AuditLog])
def read_audit_logs(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """Get all audit logs with pagination"""
    audit_logs = get_audit_logs(db, skip=skip, limit=limit)
    return audit_logs

@router.get("/entity/{entity_type}/{entity_id}", response_model=List[AuditLog])
def read_audit_logs_by_entity(
    entity_type: str,
    entity_id: int,
    db: Session = Depends(get_db)
):
    """Get audit logs for a specific entity"""
    audit_logs = get_audit_logs_by_entity(
        db, entity_type=entity_type, entity_id=entity_id
    )
    return audit_logs

@router.post("/", response_model=AuditLog, status_code=status.HTTP_201_CREATED)
def create_new_audit_log(
    audit_log: AuditLogCreate,
    db: Session = Depends(get_db)
):
    """Create a new audit log entry"""
    return create_audit_log(db=db, audit_log=audit_log)

@router.get("/deliverable/{deliverable_id}", response_model=List[AuditLog])
def read_deliverable_audit_logs(
    deliverable_id: int,
    db: Session = Depends(get_db)
):
    """Get audit logs for a specific deliverable"""
    return read_audit_logs_by_entity("deliverable", deliverable_id, db)

@router.get("/sprint/{sprint_id}", response_model=List[AuditLog])
def read_sprint_audit_logs(
    sprint_id: int,
    db: Session = Depends(get_db)
):
    """Get audit logs for a specific sprint"""
    return read_audit_logs_by_entity("sprint", sprint_id, db)

@router.get("/signoff/{signoff_id}", response_model=List[AuditLog])
def read_signoff_audit_logs(
    signoff_id: int,
    db: Session = Depends(get_db)
):
    """Get audit logs for a specific signoff"""
    return read_audit_logs_by_entity("signoff", signoff_id, db)

File: backend/node-backend/src/routes/audit.js
const express = require('express');
const router = express.Router();
const { AuditLog } = require('../models');

/**
 * @route GET /api/audit
 * @desc Get all audit logs with pagination
 * @access Public
 */
router.get('/', async (req, res) => {
  try {
    const { skip = 0, limit = 100 } = req.query;
    const auditLogs = await AuditLog.findAll({
      offset: parseInt(skip),
      limit: parseInt(limit),
      order: [['created_at', 'DESC']]
    });
    
    res.json(auditLogs);
  } catch (error) {
    console.error('Error fetching audit logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/audit/entity/:entityType/:entityId
 * @desc Get audit logs for a specific entity
 * @access Public
 */
router.get('/entity/:entityType/:entityId', async (req, res) => {
  try {
    const { entityType, entityId } = req.params;
    
    const auditLogs = await AuditLog.findAll({
      where: {
        entity_type: entityType,
        entity_id: parseInt(entityId)
      },
      order: [['created_at', 'DESC']]
    });
    
    res.json(auditLogs);
  } catch (error) {
    console.error('Error fetching entity audit logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route POST /api/audit
 * @desc Create a new audit log entry
 * @access Private
 */
router.post('/', async (req, res) => {
  try {
    const auditLogData = req.body;
    
    const auditLog = await AuditLog.create(auditLogData);
    
    res.status(201).json(auditLog);
  } catch (error) {
    console.error('Error creating audit log:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/audit/deliverable/:deliverableId
 * @desc Get audit logs for a specific deliverable
 * @access Public
 */
router.get('/deliverable/:deliverableId', async (req, res) => {
  try {
    const { deliverableId } = req.params;
    
    const auditLogs = await AuditLog.findAll({
      where: {
        entity_type: 'deliverable',
        entity_id: parseInt(deliverableId)
      },
      order: [['created_at', 'DESC']]
    });
    
    res.json(auditLogs);
  } catch (error) {
    console.error('Error fetching deliverable audit logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/audit/sprint/:sprintId
 * @desc Get audit logs for a specific sprint
 * @access Public
 */
router.get('/sprint/:sprintId', async (req, res) => {
  try {
    const { sprintId } = req.params;
    
    const auditLogs = await AuditLog.findAll({
      where: {
        entity_type: 'sprint',
        entity_id: parseInt(sprintId)
      },
      order: [['created_at', 'DESC']]
    });
    
    res.json(auditLogs);
  } catch (error) {
    console.error('Error fetching sprint audit logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/audit/signoff/:signoffId
 * @desc Get audit logs for a specific signoff
 * @access Public
 */
router.get('/signoff/:signoffId', async (req, res) => {
  try {
    const { signoffId } = req.params;
    
    const auditLogs = await AuditLog.findAll({
      where: {
        entity_type: 'signoff',
        entity_id: parseInt(signoffId)
      },
      order: [['created_at', 'DESC']]
    });
    
    res.json(auditLogs);
  } catch (error) {
    console.error('Error fetching signoff audit logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;